### PR TITLE
Add ArraySegment Enumerator Tests

### PR DESCRIPTION
**Pull request by AlexRadch**
Add tests for ArraySegment Enumerator. See https://github.com/dotnet/corefx/issues/14170

### CODE REVIEW

**Comment by AlexRadch** [comment]
@shmao, @zhenlan I have strange serialization error with serialization tests can you help me?

`/mnt/resource/j/workspace/dotnet_corefx/master/centos7.1_debug_prtest/src/Common/tests/System/Collections/IEnumerable.Generic.Serialization.Tests.cs(9,36): error CS0234: The type or namespace name 'Formatters' does not exist in the namespace 'System.Runtime.Serialization' (are you missing an assembly reference?) [/mnt/resource/j/workspace/dotnet_corefx/master/centos7.1_debug_prtest/src/System.Runtime/tests/System.Runtime.Tests.csproj]`

What should I add ?

**Comment by shmao** [comment]
It's most likely that your the test project need to reference `System.Runtime.Serialization.Formatters` in project.json.

**Comment by joperezr** [comment]
The reason this is failing is because you modified System.Runtime.Tests.csproj to include file `IEnumerable.Generic.Serialization.Tests.cs` which depends on System.Runtime.Serialization.Formatters.Binary and that is only available for `netstandard1.7`. We are x-compiling this test project for `netstandard1.5` and that is why you are seeing this error. The solution would be to condition the addition of that test file so that it only gets includded when building for `netstandard1.7`. To fix that, simply move that line to the section where we include things for the default targetGroup like:
```xml
  <ItemGroup Condition="'$(TargetGroup)' == ''">
    (...)
    <Compile Include="..\..\Common\tests\System\Collections\IEnumerable.Generic.Serialization.Tests.cs">
      <Link>Common\System\Collections\Generic\IEnumerable.Generic.Serialization.Tests.cs</Link>
    </Compile>
    (...)
  </ItemGroup>
```

**Comment by AlexRadch** [comment]
@dnlharvey, @AtsushiKan What is wrong with `System.Reflection.Tests.ModuleTests.ResolveTypeFail` test?
I did not change it
```
            System.Reflection.Tests.ModuleTests.ResolveTypeFail(token: 33555511) [FAIL]
       The target "BuildAllProjects" listed in a BeforeTargets attribute at "C:\mount\corefx\Tools\versioning.targets (218,11)" does not exist in the project, and will be ignored.
       RunTestsForProject:
               Assert.Throws() Failure
               Expected: typeof(System.ArgumentException)
               Actual:   (No exception was thrown)
               Stack Trace:
```

**Comment by AlexRadch** [comment]
@karelz can you tell who can help me with strange `System.Reflection.Tests.ModuleTests.ResolveTypeFail' failed test? I updated ArraySegment tests and all them is passed. Why ResolveTypeFail failed?

**Comment by AlexGhiondea** [comment]
@AlexRadch it looks like the test which expected to fail for a bad token is now able to resolve.

@alexperovich does this sound familiar?

**Comment by karelz** [comment]
@AlexRadch did you get the test failure under debugger? Does it repro consistently or is it intermittent failure?

**Comment by alexperovich** [comment]
Yea, I added those tests. It tries to use Module.ResolveType on "bad" metadata token values. One is a method metadata token which will always be bad. Something made the value 33555511 resolvable as a type token.

**Comment by AlexRadch** [comment]
@alexperovich How I can change this test to pass it?

**Comment by alexperovich** [comment]
@AlexRadch change the test input that is failing. Here: https://github.com/dotnet/corefx/blob/master/src/System.Runtime/tests/System/Reflection/ModuleTests.cs#L186.
My guess is that the third entry is the failing one. The 1000 might need to be increased.

**Comment by AlexGhiondea** [comment]
@alexperovich if we want a bad token number, why not pick a random number instead?

**Comment by AlexRadch** [comment]
@AlexGhiondea I think that random number sometimes will be correct.

**Comment by AlexRadch** [comment]
@AlexGhiondea @joperezr I fixed BadResolveTypes issue. Code review please.

**Comment by joperezr** [comment]
@dotnet-bot test Innerloop OSX Debug Build and Test please

**Comment by karelz** [comment]
@alexperovich do you understand why @AlexRadch's change triggered the test to fail consistently?
If it is unrelated, the test fix should not be part of this PR. If it is related, we need to fully understand why.
Marking as blocked ...

**Comment by AlexRadch** [comment]
@karelz I changed `System.Reflection.Tests.ModuleTests.ResolveTypeFail` test to fix issue with it.

@alexperovich answer
> @AlexRadch change the test input that is failing. Here: https://github.com/dotnet/corefx/blob/master/src/System.Runtime/tests/System/Reflection/ModuleTests.cs#L186.
My guess is that the third entry is the failing one. The 1000 might need to be increased.

So I only changed 1000 to 10000 in `ResolveTypeFail` test to pass it as @alexperovich answered. So I think PR should be unblocked.

Also I want to write many other tests to my PRs and all of them will use changes from this PR so I am waiting while this PR will be merged.

What I should change to merge this PR without blocking?

Before it was not merged because it not pass `System.Reflection.Tests.ModuleTests.ResolveTypeFail` test that have issue. After I changed issue in `System.Reflection.Tests.ModuleTests.ResolveTypeFail` test. This PR is blocked.

So now I do not understand how change this PR to merge it.

**Comment by karelz** [comment]
> @AlexRadch I changed System.Reflection.Tests.ModuleTests.ResolveTypeFail test to fix issue with it.

I know you did. It, However, it does not explain why it failed in the first place with your changes. That's the key thing missing - we are blocking on that understanding.

**Comment by AlexRadch** [comment]
@karelz I see @alexperovich answer that 1000 is too small and should be increased. That I did.

>@AlexRadch change the test input that is failing. Here: https://github.com/dotnet/corefx/blob/master/src/System.Runtime/tests/System/Reflection/ModuleTests.cs#L186.
My guess is that the third entry is the failing one. The 1000 might need to be increased.

I think it because I added serialization tests to `System.Runtime.Tests.csproj` see https://github.com/dotnet/corefx/pull/14431/files#diff-6223985b207d5111dc7d0c2a37cff4e7R142

I do not know why but  `System.Runtime.Tests` did not test any collection on serialization. So 1000 was enough for reflection. Now when I included serialization tests, it is not enough. 

To unblock this PR,  should I remove serialization tests from `System.Runtime.Tests.csproj` to pass old `ResolveTypeFail` test with 1000?

**Comment by karelz** [comment]
Ah, I missed the fact that we added more tests into the module - that could have made the +1000 a valid token. That would explain things sufficiently. Removing the blocking status for now ...

@alexperovich can you please confirm?

**Comment by alexperovich** [comment]
Yes, that is correct. When we add more types into the module that could make +1000 a valid token.

**Comment by AlexGhiondea** [comment]
@alexperovich surely there is a way to fix the test so that we never have to worry about not getting an invalid token when new types are added! 

Can you work with @AlexRadch to figure out such a way?

**Comment by karelz** [comment]
Agreed. We should file a separate issue to improve the test. IMO we should not block this PR on it.

**Comment by AlexGhiondea** [comment]
Yup -- let's not block on this. 

Filed https://github.com/dotnet/corefx/issues/14625 to track it.

**Comment by AlexRadch** [comment]
@AlexGhiondea @joperezr joperezr PR is unblocked. Code review please.

File: src/System.Runtime/tests/System/ArraySegmentTests.cs [hunk 1]
```diff
@@ -5,64 +5,115 @@
...
+        #endregion
+
+        #region List<T> Helper Methods
+
+        protected virtual ArraySegment<T> Factory()
+        {
+            return new ArraySegment<T>();
+        }
+
+        protected virtual ArraySegment<T> Factory(int count, int offset, int lenght)
```
Comment (AlexGhiondea) [inline src/System.Runtime/tests/System/ArraySegmentTests.cs]: nit: typo: `lenght` -> `length`

**Comment by karelz** [comment]
Thanks @AlexRadch for your contribution!

**Comment by AlexRadch** [comment]
Thanks to all!

**Comment by jamesqo** [comment]
@karelz Do you know what this PR actually did? I took 2 quick glances but I'm not sure what it's purpose was. It didn't expose the `Enumerator ArraySegment<T>.GetEnumerator()` API for https://github.com/dotnet/corefx/issues/14170, because it's not there in the ref file.

**Comment by karelz** [comment]
@jamesqo I think it added just bunch of tests.

**Pull request merged**